language-support/ts/daml-ledger/index.test.ts

```diff
@@ -0,0 +1,232 @@
+    stream.on("change", mockChange);
+
+    mockInstance.serverOpen();
+    expect(mockSend).toHaveBeenCalledTimes(1);
+    expect(mockSend).toHaveBeenLastCalledWith({templateIds: [Foo.templateId]});
+    const restoreConsole = mockConsole();
+    mockInstance.serverSend('mickey mouse');
+    expect(console.error).toHaveBeenCalledWith("Ledger.streamQuery unknown message", "mickey mouse");
+    restoreConsole();
+  });
+
+  test("receive warnings", () => {
+    const ledger = new Ledger(mockOptions);
+    const stream = ledger.streamQuery(Foo);
+    stream.on("change", mockChange);
+    const restoreConsole = mockConsole();
+    mockInstance.serverSend({ warnings: ["oh oh"] });
+    expect(console.warn).toHaveBeenCalledWith("Ledger.streamQuery warnings", {"warnings": ["oh oh"]});
+    restoreConsole();
+  });
+
+  test("receive errors", () => {
+    const ledger = new Ledger(mockOptions);
+    const stream = ledger.streamQuery(Foo);
+    stream.on("change", mockChange);
+    const restoreConsole = mockConsole();
+    mockInstance.serverSend({ errors: ["not good!"] });
+    expect(console.error).toHaveBeenCalledWith("Ledger.streamQuery errors", { errors: ["not good!"] });
+    restoreConsole();
+  });
+
+  test("receive empty events", () => {
+    const ledger = new Ledger(mockOptions);
+    const stream = ledger.streamQuery(Foo);
+    stream.on("change", state => mockChange(state));
+    mockInstance.serverSend({ events: [] });
+    expect(mockChange).toHaveBeenCalledTimes(1);
+    expect(mockChange).toHaveBeenLastCalledWith([]);
+  });
+
+  test("receive one event", () => {
+    const ledger = new Ledger(mockOptions);
+    const stream = ledger.streamQuery(Foo);
+    stream.on("change", state => mockChange(state));
+    mockInstance.serverSend({ events: [fooEvent(1)] });
+    expect(mockChange).toHaveBeenCalledTimes(1);
+    expect(mockChange).toHaveBeenLastCalledWith([fooCreateEvent(1)]);
+  });
+
+  test("receive several events", () => {
+    const ledger = new Ledger(mockOptions);
+    const stream = ledger.streamQuery(Foo);
+    stream.on("change", state => mockChange(state));
+    mockInstance.serverSend({ events: [1, 2, 3].map(fooEvent) });
+    expect(mockChange).toHaveBeenCalledTimes(1);
+    expect(mockChange).toHaveBeenCalledWith([1, 2, 3].map(fooCreateEvent));
+  });
+
+  test("drop matching created and archived events", () => {
+    const ledger = new Ledger(mockOptions);
+    const stream = ledger.streamQuery(Foo);
+    stream.on("change", state => mockChange(state));
+    mockInstance.serverSend({ events: [fooEvent(1), fooEvent(2)] });
+    expect(mockChange).toHaveBeenCalledTimes(1);
+    expect(mockChange).toHaveBeenCalledWith([fooCreateEvent(1), fooCreateEvent(2)]);
+    mockChange.mockClear();
+    mockInstance.serverSend({ events: [fooArchiveEvent(1)]});
+    expect(mockChange).toHaveBeenCalledTimes(1);
+    expect(mockChange).toHaveBeenCalledWith([fooCreateEvent(2)]);
+  });
+});
+
+describe("streamFetchByKey", () => {
```

Comment:
Testing `streamFetchByKey` on a template without a key type doesn't really make sense. Let's fix the in a followup PR.

Reply:
I added a key.